### ISSUE TITLE
Tikz graphics need attention

### ISSUE THREAD

**Issue by Abhi-1U** [Abhi-1U/texor#7]
### The Issue
Currently Tikz graphics is not being included in the generated files.
I have observed that rmarkdown via knittr tries to compile LaTeX and generate something but currently fails at it.
 

### Metadata about software versions and System

Linux Mint 20.3 X86_64
R version 3.6.3 (2020-02-29) -- "Holding the Windsock"
RStudio 2022.02.1 Build 461
pandoc 2.15
pdfTeX 3.14159265-2.6-1.40.20 (TeX Live 2019/Debian)

**Comment by Abhi-1U** [comment]
Its weird when you have multiple ways to do something but in the end both are half baked ways, for instance in including tikz code there are many ways:

1. Use R-code to read tikz section, isolate it, compile it independently and then include the generated image back as raster graphics.
2. Use R-code to read tikz section and meta data(label and caption), store it in seperate text files, then use a pandoc filter to pre process latex file by reading raw_latex and replacing the tikz section with a custom environment which translates to a div in the AST during its next transformation. This div is then treated by a lua filter which  reads the text files and pieces back the raw_tex code. Then when R-markdown reads it tries to isolate it and compile it in its own way, where if some author uses a legacy/custom tikz library which is not available, the whole process breaks.
3. Use a combination of R-code, lua-filters,pre-processing the latex file and then converting it again, for a single feature which faces the next problem of repeating them for multiple tikz images.
4. At this point, I fear that manipulating the tex file would actually break it and in my experience using R to manipulate the text has rendered the file corrupted sometimes and pandoc just generates empty output even though the file is almost identical.
5. Sometimes I feel like to modify pandoc itself and customize the latex reader to read tikz, but this is a very tough and out of scope work.

**Comment by cderv** [comment]
> but currently fails at it.

What is the failure exactly ?

**Comment by Abhi-1U** [comment]
> What is the failure exactly ?

Actually it was with an older version of my package and configuration errors, currently I have fixed it and that does not happen now

> 4. At this point, I fear that manipulating the tex file would actually break it and in my experience using R to manipulate the text has rendered the file corrupted sometimes and pandoc just generates empty output even though the file is almost identical. 

I had a big revelation today I had been manipulating source files using `setdiff(source_file_content, unnecessary_lines)` and unintentionally  latex closing brackets `}` were removed, that was the reason pandoc could not compile the document. Hence my apprehension with using R to manipulate files has eased off.

**Comment by Abhi-1U** [comment]
After a long journey here we are, as I complete the final piece of the puzzle I want to highlight how tikz/algorithm images are converted. 

## The process of conversion

1. tikz images are easy to extract using regex or pandoc filter in (latex+raw_tex mode), but they have an associated style set usually above the figure. Now to pull that piece I had to write a custom function to [pre_process](https://github.com/Abhi-1U/texor/blob/afd49cc1161b37065a94bd041fbf41c2399082a3/R/tikz-tools.R#L126-#L149) the tikz image using [pandoc lua filter](https://github.com/Abhi-1U/texor/blob/afd49cc1161b37065a94bd041fbf41c2399082a3/inst/extract_tikz.lua) and extract the style set into a YAML type file with custom blocks and identifier numbers. 
 The YAML type file looks something like this : 
```
image: 1
\tikzset{
    dot diameter/.store in=\dot@diameter,
    dot diameter=2pt,
    dot spacing/.store in=\dot@spacing,
    dot spacing=4.5pt,
    dots/.style={
        line width=\dot@diameter,
        line cap=round,
        dash pattern=on 0pt off \dot@spacing
    }
}
image-end: 1
```
This makes extracting the style set outside the figure block possible.

2. Next a R function searches RJwrapper.tex for the tikzlibs and tries to include them. while another R function extracts the figure and identifies it as tikz.
3. This is where my figure object model comes to play. This is a way to create a metadata object about all figures to document small details about each figure like image data, type, extension, path and a whole load of extra metadata which are also specific to tikz/algorithm images.
The object stores data in a similar structure to this : 

```
$image_number
[1] 1

$doc_start_line
[1] 201

$doc_end_line
[1] 229

$data
 [1] "\\begin{figure}[h!]"
# ... tikz image data                                                                                                           
[29] "\\end{figure}"                                                                                                                   

$istikz
[1] TRUE

$isalgorithm
[1] FALSE

$tikzlib
[1] "\\usetikzlibrary{decorations.pathreplacing}"

$image_count
[1] 1

$caption
[1] "a caption"

$label
[1] "fig:TimelineExample"

$tikzstyle
 [1] "\\tikzset{"                                    "    dot diameter/.store in=\\dot@diameter,"   
 [3] "    dot diameter=2pt,"                         "    dot spacing/.store in=\\dot@spacing,"     
 [5] "    dot spacing=4.5pt,"                        "    dots/.style={"                            
 [7] "        line width=\\dot@diameter,"            "        line cap=round,"                      
 [9] "        dash pattern=on 0pt off \\dot@spacing" "    }"                                        
[11] "}"                                            

$extension
[1] "png"
```

  4.   After piecing the different aspects of the tikz image, an R function assembles this into a standalone LaTeX file to compile the image to PDF, then convert the PDF to PNG to be included in the article. 
 

## why reinvent the wheel

#### 1. There are R packages like [texPreview](https://yonicd.github.io/texPreview/) then why would I do the last part of conversion by traditional methods?  

Ans: I think the reason would be that I take time to adapt to using different libraries. I write code in the manner where I express my logic and R is just a language that provides the tools to achieve that. Hence, there is a tendency to reinvent the wheel, implementing things on my own. If I wasn't familiar with pandoc, then I might have gone on to create my own document conversion tool.  On a positive note though, I will try out the tex preview and implement it soon. 
 
####  2. Why don't you use[ R-markdown block construct ](https://github.com/yihui/knitr-examples/blob/master/058-engine-tikz.Rmd) for tikz images ?
Ans: There are two reasons for that, 
1. It does not support an algorithm environment. hence not an ideal solution for that.
2. If I would have used R-chunk for tikz images only, then Figure numbering would not have been uniform. Because other images are numbered by a Lua filter within the caption, hence a possibility that R-chunk numbering could have been different and the numbering would have been irregular.

**Comment by Abhi-1U** [comment]
Fixed Issue #7 and #32 in this commit 96441a4af1c6e428b312ed954ac5487981e6fa67 .
However one must note, that this solution is not a foolproof solution and it could break sometimes
